nntrainer/utils/profiler.cpp

```diff
@@ -0,0 +1,220 @@
+// SPDX-License-Identifier: Apache-2.0
+/**
+ * Copyright (C) 2020 Jihoon Lee <jhoon.it.lee@samsung.com>
+ *
+ * @file   profiler.cpp
+ * @date   09 December 2020
+ * @brief  Profiler related codes to be used to benchmark things
+ * @see    https://github.com/nnstreamer/nntrainer
+ * @author Jihoon Lee <jhoon.it.lee@samsung.com>
+ * @bug    No known bugs except for NYI items
+ *
+ */
+#include <algorithm>
+#include <iomanip>
+#include <numeric>
+#include <sstream>
+
+#include <nntrainer_log.h>
+#include <profiler.h>
+
+namespace nntrainer {
+namespace profile {
+
+ProfileListener::ProfileListener(Profiler *profiler_, std::vector<int> events) :
+  profiler(profiler_) {
+  if (profiler != nullptr) {
+    profiler->subscribe(this, events);
+  }
+}
+
+ProfileListener::~ProfileListener() noexcept {
+  if (profiler != nullptr) {
+    try {
+      profiler->unsubscribe(this);
+    } catch (...) {
+      ml_logw("unsubscribing profiler failed may cause undefined error");
+    }
+  }
+}
+
+void GenericProfileListener::onNotify(const int event,
+                                      const std::chrono::milliseconds &value) {
+
+  auto iter = time_taken.find(event);
+
+  if (iter == time_taken.end()) {
+    time_taken[event] = std::vector<std::chrono::milliseconds>{};
```

Comment:
can we added this at creation? this is also in critical path.

Reply:
Done